### PR TITLE
Add incremental update

### PR DESCRIPTION
**Pull request by Guillemdb**
Signed-off-by: src-d <src-d@sourced.tech>

### CODE REVIEW

**Comment by Guillemdb** [comment]
Closes https://github.com/src-d/parquetry/issues/2

**Comment by Guillemdb** [comment]
If you like the structure I will add type hinting, comments and tests, but I wanted to be sure the code make sense first.

File: parquet2sql/parquet2sql.py [hunk 1]
```diff
@@ -3,7 +3,9 @@
 import argparse
 import logging
 
+import numpy
 import pandas as pd
+import sqlalchemy
 from sqlalchemy import create_engine
```
Comment (vmarkovtsev) [inline parquet2sql/parquet2sql.py]:
```suggestion
```
Let's kill this import then

File: parquet2sql/parquet2sql.py [hunk 2]
```diff
@@ -22,9 +24,48 @@ def get_parser():
...
+
+    def __init__(self, db):
+        self.db = db
+        self.engine = create_engine(db)
+
+    def process_parquet(self, parquet, table):
+        df = self._load_parquet(parquet)
+        self._incremental_delete(df, table)
+        self._incremental_update(df, table)
+        self.log.info("File %s saved to table %s at %s", parquet, table, self.db)
```
Comment (vmarkovtsev) [inline parquet2sql/parquet2sql.py]:
```suggestion
        self.log.info("File %s was saved to table %s at %s", parquet, table, self.db)
```

File: parquet2sql/parquet2sql.py [hunk 3]
```diff
@@ -22,9 +24,48 @@ def get_parser():
...
+    def _load_parquet(self, parquet):
+        df = pd.read_parquet(parquet)
+        return df
+
+    def _incremental_delete(self, df, table_name):
+        def filter_data(row, table):
+            conds = [table.c[col] == row[col] for col in table.columns.keys()]
+            return sqlalchemy.and_(*conds)
+        if "op" not in df.columns:
+            raise KeyError("`op` not in dataframe columns.")
```
Comment (vmarkovtsev) [inline parquet2sql/parquet2sql.py]:
```suggestion
            raise KeyError("There must be an `op` column in the Parquet table.")
```

File: parquet2sql/parquet2sql.py [hunk 4]
```diff
@@ -22,9 +24,48 @@ def get_parser():
...
+        delete_condtion = delete_df.apply(lambda x: filter_data(x, table), axis=1).values.tolist()
+        delete_condtion = sqlalchemy.or_(*delete_condtion)
+        # Execute delete
+        delete = table.delete().where(delete_condtion)
+        with self.engine.connect() as conn:
+            conn.execute(delete)
+
+    def _incremental_update(self, df, table):
+        if "op" not in df.columns:
+            raise KeyError("`op` not in dataframe columns.")
```
Comment (vmarkovtsev) [inline parquet2sql/parquet2sql.py]:
```suggestion
            raise KeyError("There must be an `op` column in the Parquet table.")
```

File: parquet2sql/parquet2sql.py [hunk 5]
```diff
@@ -22,9 +24,48 @@ def get_parser():
...
+        add_data = df.loc[df["op"]].drop("op", axis=1)
+        add_data.to_sql(table, self.engine, if_exists='append', index=False)
```
Comment (vmarkovtsev) [inline parquet2sql/parquet2sql.py]:
```suggestion
        add_data.to_sql(table, self.engine, if_exists="append", index=False)
```

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

**Review by vmarkovtsev — APPROVED** [review]

**Comment by vmarkovtsev** [comment]
@Guillemdb Please bump the version and I will push the new tag.

**Pull request merged**